Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+
+The current (StatefulSet) deployment brings along the corresponding pitfalls:
+* Load management & scalability: Since one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster, users would need to calculate/estimate the load and scalability of the whole cluster to decide on replicas and sharding strategies. Estimating cluster-wide load and scalability is a much harder task than estimating node-wide load and scalability. Even though they can use helping tools like Horizontal Pod Autoscaler (HPA), that's still additional complexity.
+* Security: Similarly, cluster-wide security is a much bigger problem than node-wide security.
+
+At the moment, the proposed DaemonSet deployment mode doesn’t mean to be a replacement for the current StatefulSet mode. It’s actually a solution for use cases where StatefulSet may not be the best choice. We’ll keep the StatefulSet mode as long as there’s user need for it.
+
+
+## 3. Audience
+
+Users with use cases where:
+* Scraped load is very large or hard to estimate.
+* Scalability is hard to predict.
+* Security is a big concern.
+
+
+## 4. Goals
+
+Provide an MVP version of the DaemonSet deployment of Prometheus Agent to the Audience.
+In specific, the MVP will need to:
+* Allow users to deploy one Prometheus Agent pod per node.
+* Allow users to restrict which set of nodes they want to deploy Prometheus Agent, if desired.
+* Allow each Prometheus Agent pod to scrape from the pods on the same node (PodMonitor support).
+
+
+## 5. Non-Goals
+
+The non-goals are the features that are not easy to implement and require more investigation. We will need to investigate whether there are actual user needs for them, if yes, then how to best implement them. We’ll handle these after the MVP.
+* ServiceMonitor support: There's a performance issue regarding this feature. Since each Prometheus Agent running on a node requires one watch, making all Prometheus Agent pods watch all endpoints will put a huge stress on Kubernetes API server. This is the main reason why GMP hasn’t supported this, even though there are user needs stated in some issues ([#362](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/362), [#192](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192)). However, as discussed with Danny from GMP [here](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192#issuecomment-2028850846), ServiceMonitor support based on EndpointSlice seems like a viable approach. We’ll investigate this further after the MVP.
+* Storage: We will need to spend time studying more about the WAL, different storage solutions provided by Kubernetes, and how to gracefully handle storage in different cases of crashes. For example, there’s an [issue in Prometheus](https://github.com/prometheus/prometheus/issues/8809) showing that samples may be lost if remote write didn’t flush cleanly. We’ll investigate these further after the MVP.
```

Comment:
What about mixed deployment cases? Daemon set vs others? Would it be part of goal or non goal?

Reply:
@simonpasquier once mentioned mixed modes cases. Do we need to do additional things if we want to enable mixed modes?

Reply:
Not sure, what matters is what intention your have here when testing/designing features. It feels mixed modes is a goal then?

Reply:
What do you mean by mixed deployment goals @bwplotka?
What I mentioned before is that someone could very well deploy "statefulset" Prometheus/PrometheusAgent resources alongside "daemonset" PrometheusAgent resources.

Reply:
> What I mentioned before is that someone could very well deploy "statefulset" Prometheus/PrometheusAgent resources alongside "daemonset" PrometheusAgent resources.

Exactly that - mixed deployment. I mean we should put that in goals section to keep allowing those cases 👍🏽

Reply:
I described this in Pitfalls of the current solution and Audience sections.

Reply:
I described this in Pitfalls of the current solution and Audience sections in the last commit. But I'm thinking whether I should put it in Goals too.

@simonpasquier @ArthurSens @kakkoyun Should we clarify in How that we won't allow switching from a live StatefulSet to DaemonSet => if users are using StatefulSet and want to deploy mixed mode, they will have to first deleting the StatefulSet object, then deploy DaemonSet, then deploy StatefulSet again?

AFAIU Goals/Non-goals are what guide the How section. If we need to clarify this in How then maybe we need to add mixed deployment in Goals then.